File: bnhsdk/ios/Classes/Extensions.swift
//
//  Extensions.swift
//  binah_flutter_sdk
//
//  Created by Giora Vered on 26/03/2023.
//

import Foundation

extension VitalSign {
    
    func toMap() -> [String: Any?]? {
        var confidence: VitalSignConfidence?
        var value: Any
        
        if let vitalSign = self as? VitalSignPulseRate {
            value = vitalSign.value
            confidence = vitalSign.confidence
        }
        else if let vitalSign = self as? VitalSignRespirationRate {
            value = vitalSign.value
            confidence = vitalSign.confidence
        }
        else if let vitalSign = self as? VitalSignSDNN {
            value = vitalSign.value
            confidence = vitalSign.confidence
        }
        else if let vitalSign = self as? VitalSignPRQ {
            value = vitalSign.value
            confidence = vitalSign.confidence
        }
        else if let vitalSign = self as? VitalSignMeanRRI {
            value = vitalSign.value
            confidence = vitalSign.confidence
        }
        else if let vitalSign = self as? VitalSignRRI {
            value = vitalSign.value.isEmpty ? [] : vitalSign.value.map { value in
                [
                    "interval": value.interval,
                    "timestamp": value.timestamp
                ]
            }
            confidence = vitalSign.confidence
        }
        else if let vitalSign = self as? VitalSignBloodPressure {
            value = [
                "systolic": vitalSign.value.systolic,
                "diastolic": vitalSign.value.diastolic
            ]
        }
        else if let vitalSign = self as? VitalSignStressLevel {
            value = vitalSign.value.rawValue
        }
        else if let vitalSign = self as? VitalSignPNSZone {
            value = vitalSign.value.rawValue
        }
        else if let vitalSign = self as? VitalSignSNSZone {
            value = vitalSign.value.rawValue
        }
        else if let vitalSign = self as? VitalSignWellnessLevel {
            value = vitalSign.value.rawValue
        }
        else if let vitalSign = self as? VitalSignInt {
            value = vitalSign.value
        }
        else if let vitalSign = self as? VitalSignDouble {
            value = vitalSign.value
        }
        else {
            return nil
        }
        
        var map = [
            "type": self.type,
            "value": value,
        ]
        
        if (confidence != nil) {
            map["confidence"] = [
                "level": confidence?.level.rawValue
            ]
        }
        
        return map
    }
}

extension CGRect {
    func toMap() -> [String: Int] {
        return [
            "left": Int.init(self.origin.x),
            "top": Int.init(self.origin.y),
            "width": Int.init(width),
            "height": Int.init(height)
        ]
    }
}

extension ImageData {
    func toMap() -> [String: Any?] {
        return [
            "width": Int.init(image.size.width),
            "height": Int.init(image.size.height),
            "roi": roi.isNull ? nil : roi.toMap(),
            "validity": imageValidity
        ]
    }
}

extension LicenseInfo {
    
    func toMap() -> [String: Any] {
        var map: [String: Any] = [
            "activation_id": licenseActivationInfo.activationID
        ]
        
        if let offlineMeasurements = licenseOfflineMeasurements {
            map["offline_measurements_total"] = offlineMeasurements.totalMeasurements
            map["offline_measurements_remaining"] = offlineMeasurements.remainingMeasurements
            map["offline_measurements_end_timestamp"] = offlineMeasurements.measurementEndTimestamp
        }
        
        return map
    }
}

extension SessionEnabledVitalSigns {
    
    func toMap() -> [String: Any] {
        return [
            "device_enabled": deviceEnabled.enabledVitalSigns,
            "measurement_mode_enabled": measurementModeEnabled.enabledVitalSigns,
            "license_enabled": licenseEnabled.enabledVitalSigns
        ]
    }
}

extension AlertData {
    
    func toMap() -> [String: Any] {
        return [
            "domain": domain,
            "code": code
        ]
    }
}

File: bnhsdk/android/src/main/kotlin/ai/binah/binah_flutter_sdk/Extensions.kt
package ai.binah.binah_flutter_sdk

import ai.binah.sdk.api.SessionEnabledVitalSigns
import ai.binah.sdk.api.images.ImageData
import ai.binah.sdk.api.license.LicenseInfo
import ai.binah.sdk.api.alerts.AlertData
import ai.binah.sdk.api.vital_signs.VitalSign
import ai.binah.sdk.api.vital_signs.VitalSignConfidence
import ai.binah.sdk.api.vital_signs.vitals.*
import android.graphics.Bitmap
import android.graphics.Rect
import android.view.TextureView

fun TextureView.drawBitmap(bitmap: Bitmap) {
    lockCanvas()?.apply {
        drawBitmap(
            bitmap,
            null,
            Rect(0, 0, width, bottom - top),
            null
        )
        unlockCanvasAndPost(this)
    }
}


fun Rect.toMap(): Map<String, Int> {
    return mapOf(
        Pair("left", this.left),
        Pair("top", this.top),
        Pair("width", this.width()),
        Pair("height", this.height())
    )
}

fun ImageData.toMap(): Map<String, Any?> {
    return mapOf(
        Pair("width", image.width),
        Pair("height", image.height),
        Pair("roi", roi?.toMap()),
        Pair("validity", imageValidity)
    )
}

fun AlertData.toMap(): Map<String, Any> {
    return mapOf(
        Pair("domain", domain),
        Pair("code", code)
    )
}

fun LicenseInfo.toMap(): Map<String, Any> {
    val map = mutableMapOf<String, Any>(
        Pair("activation_id", licenseActivationInfo.activationID)
    )
    licenseOfflineMeasurements?.let { offlineMeasurements ->
        map["offline_measurements_total"] = offlineMeasurements.totalMeasurements
        map["offline_measurements_remaining"] = offlineMeasurements.remainingMeasurements
        map["offline_measurements_end_timestamp"] = offlineMeasurements.measurementEndTimestamp
    }

    return map
}

fun SessionEnabledVitalSigns.toMap(): Map<String, Long> {
    return mapOf(
        Pair("device_enabled", deviceEnabledVitalSigns.enabledVitalsSigns),
        Pair("measurement_mode_enabled", measurementModeEnabledVitalSigns.enabledVitalsSigns),
        Pair("license_enabled", licenseEnabledVitalSigns.enabledVitalsSigns)
    )
}

fun VitalSign.toMap(): Map<String, Any>? {
    var confidence: VitalSignConfidence? = null
    val resolvedValue = when (this) {
        is VitalSignPulseRate -> {
            confidence = this.confidence
            value
        }
        is VitalSignRespirationRate -> {
            confidence = this.confidence
            value
        }
        is VitalSignPRQ -> {
            confidence = this.confidence
            value
        }
        is VitalSignMeanRRI -> {
            confidence = this.confidence
            value
        }
        is VitalSignSDNN -> {
            confidence = this.confidence
            value
        }
        is VitalSignRRI -> {
            confidence = this.confidence
            if (value.isEmpty()) null else value.map { rriValue ->
                mapOf(
                    Pair("interval", rriValue.interval),
                    Pair("timestamp", rriValue.timestamp)
                )
            }
        }
        is VitalSignStressLevel -> value.ordinal
        is VitalSignPNSZone -> value.ordinal
        is VitalSignSNSZone -> value.ordinal
        is VitalSignWellnessLevel -> value.ordinal
        is VitalSignBloodPressure -> mapOf(
            Pair("systolic", value.systolic),
            Pair("diastolic", value.diastolic)
        )
        else -> value
    }

    return mutableMapOf(
        Pair("type", type),
        Pair("value", resolvedValue ?: return null)
    ).apply {
        if (confidence != null) {
            put("confidence", mapOf(Pair("level", confidence.level.ordinal)))
        }
    }
}